### PR TITLE
Fix relative jumps

### LINKED ISSUE #41: JUMPS instruction conditions
[ESP32 TRM](https://www.espressif.com/sites/default/files/documentation/esp32_technical_reference_manual_en.pdf) prior to v3.6 incorrectly described conditions used by JUMPS instruction.

TRM is now updated and this has been fixed in binutils-esp32ulp: [diffs](https://github.com/espressif/binutils-esp32ulp/compare/b7d6d00246af5ed99a24b8be81868874c7c5cf4e...master), [docs](https://github.com/espressif/esp-idf/commit/1bd7d404feec85ede416edde4678ee4b65cbe61f).

Would be great to propagate these updates here as well.

### PR DESCRIPTION
**Pull request by wnienhaus**
This PR is meant to fix #41 .

The PR is more of a prototype for now (not finished). I raised it to be able to:
1. discuss the implementation of "emit multiple instructions"
2. get ideas on a behaviour of binutils-esp32ulp that I cannot yet explain.

---
**1. opcode -> Multiple instructions**
For now the implementation (see commit fa09eb3) works by the `i_jumpr` and `i_jumpr` methods returning a `tuple` when multiple instructions should be emitted, and the assembler testing for "is this a tuple" and if so, it looping over the elements. Otherwise it does what it did before, i.e. adds 1 instruction to the section. The current structure for this logic feels a bit rough, but for now I intend mostly to get feedback on the approach.

To help this, for pass 1, when we're not yet calling the `i_*` methods, because we only want to determine section sizes and label offsets, I added a function that can return "how many instructions will this opcode result in" based on the opcode and its arguments. The default return value is 1, but for the JUMPS and JUMPR opcodes and with the specific conditions that will result in 2 instructions, the function returns 2. Simple and effective, but technically a "duplication of logic" (because `i_jumps`/`i_jumpr` methods also make this decision). Perhaps there are other approaches for how to make this even simpler or more elegant?

---
**2. Weird binutils-esp32 behaviour with negative immediate offsets**
For a reason I cannot yet figure out, when using negative immediate offsets for JUMPS or JUMPR instructions, binutils-esp32 generates instructions differently than when using label references that are before the current PC (back-jumps).

The behaviour is as follows: for SYM references, if the offset is negative, set the sign bit and use the absolute value of the offset. for IMM references however, the sign bit is set the same way, but for the offset the negative (not absolute) value is used.

At the "field" level in the opcode the difference is such that:
* For a negative symbol reference, e.g. -9, sign bit is 1 and the offset is 9
* For a negative immediate reference, e.g. -2, the sign bit is 1 and the offset is 126 (which is the 7-bit two's complement of 2)

See commit 3ae1e20 for the implementation that mirrors what binutils-esp32ulp does. So "it works", in that py-esp32-ulp will now generate the same binary output, but I still don't know **why** binutils-esp32ulp treats these offsets differently (even after a fair bit of scanning the binutils-esp32ulp source code). If we need to jump 4 words back, because there is a label 4 instructions earlier, of whether I give an immediate offset of -4, I would expect the resulting jump instruction to be the same. I have not tested this on the actual device for how the ULP interprets these different "representations".

---
**3. One more thing**
One last thing, which is not yet in this PR, is another weird behaviour of binutils-esp32ulp. This relates to using constants defined with `.set` as arguments to the JUMPS and JUMPR opcodes. I would expect them to behave like immediate offsets, but they are always "interpreted" as 0.

So the following two JUMPS statements result in the exact same instruction (namely `2a00 0084`):
```
  .set const, 42
  jumps 0, 1, lt
  jumps const, 1, lt
```

I started wondering whether this is a bug in binutils-esp32ulp? Or I have wondered if "resolving the constant" is deferred by binutils-esp32ulp to the linker perhaps, and maybe we're calling the linker wrong in the compat test scripts, but so far I cannot see that this is true. The binutils-esp32ulp test scripts call the linker the same way (then again, sadly they don't have this specific example in their tests).

---
Happy to get feedback/thought/ideas on the above, and the current implementation.

### CODE REVIEW

**Comment by wnienhaus** [comment]
Nevermind about point 3 above. I just got it! It is in fact the linker that fills in the `const` and therefore the symbol must be exported with `.global const`. When `const` is exported, both py-esp32-ulp and binutils produce the same output. I think I will leave this one as-is then, because technically, with binutils-esp32ulp, it's a case of "using it wrong", if one doesn't export the symbol.

**Comment by wnienhaus** [comment]
I'm now relatively sure, that point 2 is a bug in binutils-esp32ulp. I dug into that code, debugged around a bit and found the place where it is likely wrong.

I compared what binutils-esp32ulp does with the ULP "macro approach" in the ESP-IDF, and there the offset is always converted to absolute and the 8th bit is set based on sign (as we do in py-esp32-ulp). I also read the documentation (ESP32 TRM), which is mentioned in issue #41, and it too states that that bit 0-6 (7bits) are always a positive number and the 8th bit determines how the ULP will interpret the offset (positive or negative). So 2 sources against 1 encode the negative offsets the same way between values from symbols and immediate values.

I have a fix suggestion for binutils-esp32ulp and will send it as PR to that project and see what they say. Then I guess we'll know more.

Until then I think i'll revert the last patch of this PR, and thus simply "leave-it-as-before" (which I currently think is the right way anyway). Then we can hopefully merge this change already before any response from espressif and if we need to change, we can change that later.

**Comment by wnienhaus** [comment]
I have now made a PR to the binutils-esp32ulp project (https://github.com/espressif/binutils-esp32ulp/pull/18). Let's see what they say.

**Comment by wnienhaus** [comment]
So I got a positive response to the PR to binutils-esp32ulp. I will wait until it's fully merged, and then I will update these tests to pass against the updated binutils-esp32ulp.

**Comment by wnienhaus** [comment]
Now that my fix to binutils-esp32ulp has been merged, I finally managed to clean up this PR.

Points 2 and 3 from the original description of this PR are resolved. Now we can focus on the implementation of the "multi-instruction" approach.

In summary: `opcodes.i_*` functions can now optionally return a tuple, in which case each element is appended as an instruction to the TEXT section. A separate `opcodes.no_of_instr` function can be consulted during pass 1 of assembly to determine how many instructions a given opcode will result in.

Happy to get feedback on the implementation.

File: esp32_ulp/opcodes.py
Comment (ThomasWaldmann) [inline esp32_ulp/opcodes.py:709]: hmm, naming seems unintuitive to me.

how about using skip_cond (instead of walk_cond), because the 2nd jump instruction is skipped?

how about using jump_cond (instead of result_cond), because that is the condition in the 2nd jump instruction that eventually really jumps to the target?

fix globally...

File: esp32_ulp/opcodes.py
Comment (ThomasWaldmann) [inline esp32_ulp/opcodes.py:668]: ... # eq == ge(threshold) but not ge(threshold+1)

**Review by ThomasWaldmann — CHANGES_REQUESTED** [review]
some nitpicks, but looks pretty good already!

File: esp32_ulp/opcodes.py
Comment (wnienhaus) [reply esp32_ulp/opcodes.py:343]: I usually prefer writing code to match my intent, and let the compiler optimise it. But in this case it appears MicroPython will not do any optimisation ([this appears to be the place](https://github.com/micropython/micropython/blob/4d9e657f0ee881f4a41093ab89ec91d03613744d/py/objint_longlong.c#L162)), so I will change this to bitwise operations and add the intent as a comment.

File: esp32_ulp/opcodes.py
Comment (wnienhaus) [reply esp32_ulp/opcodes.py:669]: good catch

File: esp32_ulp/opcodes.py
Comment (wnienhaus) [reply esp32_ulp/opcodes.py:709]: Very good point. I actually didn't think about this too much - I simply used the names that espressif used in binutils-esp32ulp. I do remember them not making much sense to me there, but didn't stop to think about changing them in our implementation :).

I really like your suggestions for names.

File: esp32_ulp/opcodes.py
Comment (wnienhaus) [reply esp32_ulp/opcodes.py:736]: Looks good. Nice and short. Loses the documentation of what args[2] is though. Will change it anyway.

**Comment by wnienhaus** [comment]
Fixed all the things you spotted. Thanks for that. Pushed latest code (amended commits).

File: esp32_ulp/opcodes.py
Comment (ThomasWaldmann) [inline esp32_ulp/opcodes.py:703]: move comment to next line.

File: esp32_ulp/opcodes.py
Comment (ThomasWaldmann) [reply esp32_ulp/opcodes.py:707]: add this comment also, please.

**Review by ThomasWaldmann — CHANGES_REQUESTED** [review]
thanks for the updates, 2 comment edits and we're done.

File: esp32_ulp/opcodes.py
Comment (wnienhaus) [reply esp32_ulp/opcodes.py:703]: Flüchtigkeitsfehler :). Fixed now.

**Review by ThomasWaldmann — APPROVED** [review]
great, let's merge...

**Pull request merged**
